### ISSUE TITLE
Static Typing Support

### ISSUE THREAD

**Issue by nickbalestra** [opentable/oc#680]
This is just to start the conversation around static typing.
Should we start considering to add static typing to OC either via TS or Flow? 
Would love to hear from @aselbie @matteofigus @mattiaerre @debopamsengupta @danrr @elboletaire

**Comment by elboletaire** [comment]
I started using flow just a weeks ago, and never used typescript.. so I cannot bring much personal information.

Although I've been reading a bit...

https://djcordhose.github.io/flow-vs-typescript/2016_hhjs.html#/
https://michalzalecki.com/typescript-vs-flow/
https://www.reddit.com/r/javascript/comments/39cere/typescript_vs_flow_results_from_our_investigation/

...apparently there are not too much differences between them.

I'd probably start using flow, because (I quote from the readings above):

- Flow will be a better choice when it comes to adding typings to the existing project.
- Flow requires nearly no "flow specific" artifacts.
- Using generic types you can choose from invariant (exact match), covariant (more special), and contravariant (more common).

**Comment by debopamsengupta** [comment]
A question
- will this also enforce component authors to write typed JS ? 

Haven't used wither Flow/TypeScript before...but it seems a strong point of TypeScript is existing JS is compatible with it..perhaps that flexibility will be useful for component authors ?

**Comment by elboletaire** [comment]
AFAIK this does not affect component authors. They can still do their code with or without typing their components.

**Comment by elboletaire** [comment]
Moreover, you can do some parts of the code typed and other parts without being typed. At least for what I've been working with, I've been able to work either ways... both with typed vars and without typed vars (even in the same file).

**Comment by debopamsengupta** [comment]
for component authors to use typed code, I guess a `oc-template-typescript` or `oc-template-flow` is more appropriate ?

**Comment by nickbalestra** [comment]
yes @debopamsengupta, the issue is to simply add static typing support to the oc codebase itself. This shouldn't change how people write components and yes, if you would like to add ts/flow you could add that to your own oc-template :)

**Comment by matteofigus** [comment]
I'm in favour inside templates for writing components, in fact this is possible with the OC templates.

What's the value we are bringing to the table for having oc itself rewritten in TS/Flow?
Personally, I always had a goal of making easy for people to contribute and improving the contribution rate over time, and I think a Typed system could undermine this:

1) Less easy to contribute - JS is always the more popular choice
2) Less easy to contribute, things like TS thigh you to use VS Code, while JS is a better choice for being more inclusive with people using any other editor. Same probably applies to Flow etc.
3) Less easy to contribute, more steps to understand how the codebase is organised and do your change/run the tests?
4) More complexity on the build pipelines to distribute as library

**Comment by ctrlplusb** [comment]
Would you consider leveraging something like [ncc](https://github.com/zeit/ncc) which can transpile a Node file with imports into a single file? It has TypeScript support baked in.

**Comment by matteofigus** [comment]
@ctrlplusb that would be nice. As you can see this conversation hasn't evolved too much in 2 years and now TS is a much more popular language. I think allowing developers to leverage types would actually be very useful. Unfortunately in this codebase we still don't use `import` statements (but we have some ES6).

I definitely would love something like that. I wasn't aware of ncc.

**Comment by debopamsengupta** [comment]
Is the `server.js` currently transpiled using webpack ? In which case, surely Webpack would support TS out of the box ? or with a plugin or two ?
